### ISSUE TITLE
Linked module package.json ignored and cached .yarn-metadata.json used instead

### ISSUE THREAD

**Issue by RandScullard** [yarnpkg/yarn#5390]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
I want to use a local (mocked) version of a published package (i2c-bus) in my main project in place of the real thing, so I use `yarn link` to link my local copy into the main project. The published version of i2c-bus has this script in its package.json:

    "install": "node-gyp rebuild"

...but my local version doesn't use node-gyp since it's just a mock. So my local version of package.json doesn't have an install script.

When I do yarn install on my main project, the package.json of my linked local copy of i2c-bus is ignored and the install script from the i2c-bus .yarn-metadata.json (in the yarn cache) is run instead. Then node-gyp fails, because it doesn't find any of the stuff node-gyp needs in my local copy of i2c-bus.

I know that it is using the .yarn-metadata.json as the source of the install script, because if I edit .yarn-metadata.json by hand and replace "node-gyp rebuild" with "echo \"BARFOO\"", my yarn install works fine.

**What is the expected behavior?**
I expect that when I have linked a local version of a package, the linked local package.json will be used and not the one from the published version of the package in the yarn cache. It seems like there could be many problems besides the one described above if the local version of package.json is ignored. 

**Please mention your node.js, yarn and operating system version.**
Node 8.9.4, Yarn 1.3.2, Windows 10

**Comment by rally25rs** [comment]
I think Yarn is simply checking the cache first and using what is there.

As a workaround, I think you can run `yarn cache clean i2c-bus` to remove the "published" one from the cache. Or, bumping the version number of your local one might do it as well.

**Comment by RandScullard** [comment]
I agree that Yarn is simply checking the cache first and using what is there. However, in the case where you are using a linked package, it's hard to see how that behavior could be considered correct.

I'm afraid that neither of your suggested workarounds is effective:

- `yarn cache clean` removes the cached copy, but the next `yarn install` immediately restores it.
- I've bumped the version number on my linked copy of i2c-bus, but yarn is still using the install script from the .yarn-metadata.json in the cache.

The only workaround so far is to manually edit the .yarn-metadata.json file.

**Comment by jonahgraham** [comment]
FWIW - this is still here and the effect can be fairly problematic. In my use case I am removing node-gyp (and natives) from my linked to project. But as the published project has `install` script set to `node-gyp rebuild` doing a `yarn install` on my dependent project leads to `node-gyp rebuild` running, which then fails because `binding.gyp not found`.
